### PR TITLE
Feature/evaluator

### LINKED ISSUE #3808: Evaluators need to store variable(tensor) in an op
When start writing an evaluator operator, I found that I have to store a global status like `total_samples` and `total_score` to calculate the error rate of current forward iteration, so that I can increase these values every forward, like `total_samples += batch_size` and `total_score += score`.

A simply thought is to add a member function `CreateOpVar` in `InferShapeContext`, and name the variable like `[operator_id]_var_name` to avoid variable name conflict.

Can we discuss about whether we need to add this to current framework?

### PR DESCRIPTION
**Pull request by dzhwinter**
fix #3808 
The previous evaluator doesn't separate compile time and runtime, and also we need to create some metric states for each evaluator.

### CODE REVIEW

File: doc/design/evaluator.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+## Evaluator Design
+
+### The Problem
+
+During training or serving, we provide the evaluation function to measure the model performance, e.g., accuracy, precision. In the operator based framework design, the data go through the network pipeline batch by batch. As a result, inside the operator, we only can calculate one minibatch metrics. We need to provide a mechanism to calculate the metrics for each N pass/batch the user wanted.
+
+### Evaluator Design
+Currently, every operation is expressed in the graph. we divide the evaluator process into three steps.
+
+1. Initialize the metric state necessary and add it into the block.
```
Comment (helinwang) [inline doc/design/evaluator.md]: metric state necessary -> metric state

File: doc/design/evaluator.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+2. Calculate the statistic of the metric state in every mini-batch. The single operator is only responsible for calculating necessary statistics for one mini-batch. For example, accuracy operator only calculate a minibatch data if run once.\
```
Comment (helinwang) [inline doc/design/evaluator.md]: remove "\"

File: doc/design/evaluator.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    """
+    Evalutor Base class.
+    """
+    def __init__(self):
+       """
+       create metric states and append to block
+       """ 
+       pass
+
+    def _clear_state(self):
```
Comment (helinwang) [inline doc/design/evaluator.md]: Maybe this should not be a part of Python.

File: doc/design/evaluator.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      """
+      clear metric states at the begin of each pass
+      """
+      pass
+
+    def _append_evalutor_op(self):
```
Comment (helinwang) [inline doc/design/evaluator.md]: Perhaps we just need to specify what the interface looks like to the user.

I think a very important rule is: **we don't make the decision unless we absolutely have to***. By deferring the decision what private methods to use for as long as possible gives us more information to make a good decision.

File: doc/design/evaluator.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+### Evaluator Design
+Currently, every operation is expressed in the graph. we divide the evaluator process into three steps.
+
+1. Initialize the metric state and add it into the block.
+
+2. Calculate the statistic of the metric state in every mini-batch. The single operator is only responsible for calculating necessary statistics for one mini-batch. For example, accuracy operator only calculate a minibatch data if run once.
+
+
+3. Merge the mini-batch statistics to form the evaluation result for multiple mini-batches. When it comes to distributed training/Multi-GPU training, aggregate the value from different devices.
```
Comment (typhoonzero) [inline doc/design/evaluator.md]: I saw the code below, it seems that we need some detailed description of how to implement evaluator operators in C++, how to save state in C++ and update them in python side.

File: python/paddle/v2/framework/evaluator.py [hunk 6]
```diff
@@ -1,59 +1,134 @@
...
-        :param label: variable name of label
-        :type label: string
+      Clear metric states at the begin of each pass/user specified batch
+      """
+        if program == None:
+            reset_program = Program()
+        else:
+            reset_program = program
+        for k, var in self._states.iteritems():
+            zeros = helper.create_tmp_variable(dtype=var.data_type)
```
Comment (Canpio) [inline python/paddle/v2/framework/evaluator.py]: `helper` can only create vars and ops in `main_program` and `startup_programe`. However, here we need to create them in the `reset_program`. So it's not correct to use `helper` here.

File: python/paddle/v2/framework/evaluator.py [hunk 7]
```diff
@@ -1,59 +1,134 @@
...
+        # else:
+        #     self._program = g_program
+
+    # def _update(self):
+    #     """
+    #     Updates the internal states througth operator
+    #   """
+    #     raise NotImplementedError()
+
+    def reset(self, executor, program=None):
```
Comment (Canpio) [inline python/paddle/v2/framework/evaluator.py]: Do not expose reset_program to the upper level. It should be created, run and destroyed in the `reset` function.

File: doc/design/evaluator.md (hunk 5 above)
Comment (dzhwinter) [reply doc/design/evaluator.md]: Sure, I will add the detail.
Currently, we have two options. 
option 1, just like the TensorFlow does, composing the low-level operators to compute metric.  If the performance is a real bottleneck, rewrite them in the c++ side as a new operator.
option 2, we use c++ operator to calculate every mini-batch metric and maintain some states in Python side. 
I'm not sure which is better now. I implement the option 2, how do you think?

File: python/paddle/v2/framework/evaluator.py (hunk 7 above)
Comment (dzhwinter) [reply python/paddle/v2/framework/evaluator.py]: Done.

File: python/paddle/v2/framework/evaluator.py (hunk 6 above)
Comment (dzhwinter) [reply python/paddle/v2/framework/evaluator.py]: Done.

File: doc/design/evaluator.md (hunk 5 above)
Comment (dzhwinter) [reply doc/design/evaluator.md]: Done.

File: python/paddle/v2/framework/evaluator.py [hunk 8]
```diff
@@ -1,59 +1,188 @@
...
+            outputs={"Out": [self._states["Total"]]})
+        block.append_op(
+            type="elementwise_add",
+            inputs={"X": [self._states["Correct"]],
+                    "Y": [correct]},
+            outputs={"Out": [self._states["Correct"]]})
+
+        return acc_out
+
+    def eval(self, executor, program=None):
```
Comment (Canpio) [inline python/paddle/v2/framework/evaluator.py]: `eval_program=None`

File: python/paddle/v2/framework/tests/test_fit_a_line.py [hunk 9]
```diff
@@ -31,6 +32,8 @@
     main_program=main_program,
     startup_program=startup_program)
 
+accuracy = evaluator.Accuracy(input=y_predict, label=y)
```
Comment (Canpio) [inline python/paddle/v2/framework/tests/test_fit_a_line.py]: `evaluator.accuracy()`

File: python/paddle/v2/framework/evaluator.py (hunk 8 above)
Comment (dzhwinter) [reply python/paddle/v2/framework/evaluator.py]: Done.

File: python/paddle/v2/framework/evaluator.py [hunk 10]
```diff
@@ -1,59 +1,188 @@
...
+        else:
+            self._main_program = g_main_program
+
+    def _update_ops(self, *args, **kwargs):
+        """
+        append update ops to the global states
+        """
+        raise NotImplementedError()
+
+    def reset(self, executor, program=None):
```
[the remark this answers is not in the document]
Comment (dzhwinter) [reply python/paddle/v2/framework/evaluator.py]: Done

File: python/paddle/v2/framework/tests/test_fit_a_line.py (hunk 9 above)
Comment (dzhwinter) [reply python/paddle/v2/framework/tests/test_fit_a_line.py]: removed
